consensus/src/dag/dag_network.rs

```diff
@@ -28,5 +39,136 @@ pub trait DAGNetworkSender: Send + Sync {
+    generator: ExponentialNumberGenerator,
+}
+
+impl Responders {
+    fn new(mut peers: Vec<Author>, initial_request_count: u32, max_request_count: u32) -> Self {
+        peers.shuffle(&mut rand::thread_rng());
+        Self {
+            peers,
+            generator: ExponentialNumberGenerator::new(initial_request_count, 2, max_request_count),
+        }
+    }
+
+    fn next_to_request(&mut self) -> Option<Vec<Author>> {
+        let count = self.generator.next().expect("should return a number");
+
+        if self.peers.is_empty() {
+            return None;
+        }
+        Some(
+            self.peers.split_off(
+                self.peers
+                    .len()
+                    .checked_sub(count as usize)
+                    .unwrap_or(0),
+            ),
+        )
+    }
+}
+
+pub struct RpcWithFallback {
+    responders: Responders,
+    message: ConsensusMsg,
+    timeout: Duration,
+
+    terminated: bool,
+    futures: Pin<
+        Box<FuturesUnordered<Pin<Box<dyn Future<Output = anyhow::Result<ConsensusMsg>> + Send>>>>,
+    >,
+    sender: Arc<dyn DAGNetworkSender>,
+    interval: Pin<Box<Interval>>,
+}
+
+impl RpcWithFallback {
+    pub fn new(
+        responders: Vec<Author>,
+        message: ConsensusMsg,
+        timeout: Duration,
+        sender: Arc<dyn DAGNetworkSender>,
+        time_service: TimeService,
+    ) -> Self {
+        Self {
+            responders: Responders::new(responders, 1, 4),
+            message,
+            timeout,
+
+            terminated: false,
+            futures: Box::pin(FuturesUnordered::new()),
+            sender,
+            interval: Box::pin(time_service.interval(timeout)),
+        }
+    }
+}
+
+impl Stream for RpcWithFallback {
+    type Item = anyhow::Result<ConsensusMsg>;
+
+    fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let send_rpc = move |sender: Arc<dyn DAGNetworkSender>, peer, message, timeout| async move {
+            sender.send_rpc(peer, message, timeout).await
+        };
+
+        let timeout = matches!(self.interval.as_mut().poll_next(cx), Poll::Ready(_));
+
+        if self.futures.is_empty() || timeout {
+            if let Some(peers) = Pin::new(&mut self.responders).next_to_request() {
+                for peer in peers {
+                    let future = Box::pin(send_rpc(
+                        self.sender.clone(),
+                        peer,
+                        self.message.clone(),
+                        self.timeout,
+                    ));
+                    self.futures.push(future);
+                }
+            } else if self.futures.is_empty() {
+                return Poll::Ready(None);
+            }
+        }
+
+        let result = futures::ready!(self.futures.as_mut().poll_next(cx));
+        Poll::Ready(result)
+    }
+}
+
+impl FusedStream for RpcWithFallback {
+    fn is_terminated(&self) -> bool {
+        self.futures.is_empty()
```

Comment:
I am confused. If it is empty you might want to send_rpc again.
Are not you supposed to check self.terminated?

Reply:
I think I need to set the terminated first and check here.

Reply:
I think it was working previously, because poll_next makes sure that `futures`` is always non-empty befor returning. If it `futures` is empty, then it means it is terminated. It won't work now because we check the futures first and return proactively.